guardian/archivehunter
# Fix/folder listing

## What does this change?
- overhauls the way that paths are presented according to the spec at https://codemill.atlassian.net/browse/GP-379
- creates a new index, which is used to store broken-down path representations so we can quickly load the "folder" list on large buckets (e.g. 1.5mil items +)
- adds a "reindex" function which can be used to generate the index initially (accessible from the Admin UI)
- updates the input lambda to keep the new index up-to-date as new files come in
- adds a "failsafe" restore function where you can type/paste a folder path and a bucket to lightbox/restore. It won't let you go over-quota or initiate an empty restore.

This became an emergency hotfix when the folder list in the current archive bucket stopped working because the loadbalancer was timing out before the backend could iterate all the items to get the folders.
A temporary fix of performing the iteration on the index instead did not help either, so we had to go the whole hog.
This means that it is currently deployed to PROD in a bare-bones form.

## How to test
- On DEV, observe that the folder list now loads much faster. On PROD, observe that the folder list now loads much _much_ faster!
- There should be no change in functionality 
- On DEV, use the S3 console to put some media into the bucket in a new subfolder. Observe that the media is present in the Browse view and so is its folder
- On DEV, go to Admin and go to the new "Failsafe Restore" page.  Enter a valid bucket name and path and hit "Check Path". You should see a summary (assuming no typos!) and Add to Lightbox become available.  Hit Add to Lightbox and the content will appear in your Lightbox (initiating a restore if necessary).  The backend process here is identical to the one in the Browse view, it's just a more basic way of requesting the same operation from the backend server.
- On DEV, go to Admin and go to the new Paths Caching page.  Observe the number of total paths.  Hit the reindex button, you should see that it was successfully started (right now there is no progress bar though)

## How can we measure success?
Faster UX. Ability to use Browse view on gnm-multimedia-ukarchive2019 again

## Have we considered potential risks?
Risk of the folder view failing if the index is not available.  This is negligible since it had failed anyway in the most important place.
